Fix cursor events on screen lock/unlock

OkamiW opened the pull request:
After the scene graph refactor, node_at_coords no longer return the lock surface after the screen is locked, thereby cursor_rebase_all in handle_surface_map won't work.

This commit make node_at_coords return lock surface when locked.

I also found that the cursor would disappear after screen is unlocked, thereby adding a cursor_rebase_all in handle_unlock.

emersion commented on sway/input/cursor.c:
```diff
@@ -63,6 +63,11 @@ struct sway_node *node_at_coords(
 	}
 
+	if (server.session_lock.lock) {
+		*surface = server.session_lock.lock->focused;
```
`focused` represents the keyboard focus, but this codepath is about the pointer. Moreover, the surface-local coordinates `sx`/`sy` are left unset.

The session lock surfaces are added to `output->layers.session_lock`, so I wonder why this doesn't work?

cc @Nefsen402

kennylevinsen commented:
Hmm, this is a bit of an event ordering issue with `wlr_scene`.

The order of events here is:
1. Surface is committed
2. Surface role commit implementations are called
3. Surface role commit handlers call `wlr_surface_map`
4. Surface role map implementations are called
5. Map event fires
6. Commit event fires
7. Scene surface helper attaches new buffer to the surface

Dealing with this would require an "initial commit" sort of commit handler... :/

Nefsen402 commented:
I've been pondering for a while if we should get rid of all the cursor rebase stuff and add some sort of callback or just go through with `wlr_scene_cursor` finally. Scene could implement cursor rebasing on its own and all these problems would go away similar to how we handled xwayland restacking.

kennylevinsen commented:
It could be useful, but there are some caveats.

One is that the effect of `cursor_rebase` is currently conditional on the ongoing seatop, with only the default seatop performing focus updates. It could also be argued that it's a bug that `cursor_rebase` doesn't also update keyboard focus when `focus_follows_mouse always` is set.

Another is that, even if scene have a helper, there should preferably still be a way for compositors to do things on their own.

Commit:
Fix cursor events on lock/unlock

See https://github.com/labwc/labwc/pull/1858

Delay refocus operations because only the role-specific surface
commit/map handler has been processed in wlroots at this moment and
node_at_coords returns the WLR_SCENE_NODE_RECT as a buffer has not
been actually attached to the surface.

OkamiW commented:
It seems that Labwc has a fix for the same problem: https://github.com/labwc/labwc/pull/1858
`wl_event_loop_add_idle` can be used here to delay the actual operations after the buffer has been attached.
This should be a proper fix for the issue.

Commit:
Fix sway_session_lock_has_surface

Before this commit, if the cursor is at screen center, and the lock is
swaylock, the cursor would be at swaylock's subsurface(the indicator).

Since it's not the lock surface, `handle_rebase` would refuse to
rebase the cursor to there. Thereby the cursor enter event won't be
sent to swaylock.

This commit fix the issue.

OkamiW commented:
Gentle ping.

OkamiW commented:
Any update on this?

emersion commented on sway/lock.c:
```diff
@@ -325,11 +332,26 @@ void sway_session_lock_add_output(struct sway_session_lock *lock,
 	struct sway_session_lock_output *lock_output;
 	wl_list_for_each(lock_output, &lock->outputs, link) {
-		if (lock_output->surface && lock_output->surface->surface == surface) {
+		if (lock_output->surface && (lock_output->surface->surface == surface
+					     || wlr_surface_has_subsurface(lock_output->surface->surface, surface))) {
```
This only handles one level of sub-surfaces. If we want to handle arbitrary nested sub-surfaces, we should use `wlr_surface_for_each_surface()`.

OkamiW replied:
The code has been updated to use `wlr_surface_for_each_surface`.

emersion commented:
In general the approach of delaying the focus updates are more of a workaround than a proper fix. However it's not clear what the proper fix could look like.
